src/core_plugins/kibana/public/dashboard/actions/embeddables.js

```diff
@@ -0,0 +1,43 @@
+import { createAction } from 'redux-actions';
+
+export const destroyEmbeddable =
+  createAction('DESTROY_EMBEDDABLE', (panelId, embeddableHandler) => {
+    if (embeddableHandler) {
+      embeddableHandler.destroy(panelId);
+    }
+    return panelId;
+  });
+
+export const embeddableRenderFinished =
+  createAction('EMBEDDABLE_RENDER_FINISHED', (panelId, embeddable) => ({ embeddable, panelId }));
+
+export const embeddableRenderError =
+  createAction('EMBEDDABLE_RENDER_ERROR', (panelId, error) => ({ panelId, error }));
+
+/**
+ *
+ * @param embeddableHandler {EmbeddableHandler}
+ * @param panelElement {Node}
+ * @param panelId {string}
+ * @param containerApi {ContainerAPI}
+ * @return {function(*, *)}
+ */
+export function renderEmbeddable(embeddableHandler, panelElement, panelId, containerApi) {
+  return (dispatch, getState) => {
+    const { dashboardState } = getState();
+    const panelState = dashboardState.panels[panelId];
```

Comment:
This isn't using selectors, so it's "locking down" the state shape. Is there a reason for not injecting it into the action instead? E.g.

```diff
diff --git a/src/core_plugins/kibana/public/dashboard/actions/embeddables.js b/src/core_plugins/kibana/public/dashboard/actions/embeddables.js
index 236cd9a79..df4d1e70f 100644
--- a/src/core_plugins/kibana/public/dashboard/actions/embeddables.js
+++ b/src/core_plugins/kibana/public/dashboard/actions/embeddables.js
@@ -22,22 +22,19 @@ export const embeddableRenderError =
  * @param containerApi {ContainerAPI}
  * @return {function(*, *)}
  */
-export function renderEmbeddable(embeddableHandler, panelElement, panelId, containerApi) {
+export function renderEmbeddable(embeddableHandler, panelState, panelElement, containerApi) {
   return (dispatch, getState) => {
-    const { dashboardState } = getState();
-    const panelState = dashboardState.panels[panelId];
-
     if (!embeddableHandler) {
-      dispatch(embeddableRenderError(panelId, new Error(`Invalid embeddable type "${panelState.type}"`)));
+      dispatch(embeddableRenderError(panelState.panelIndex, new Error(`Invalid embeddable type "${panelState.type}"`)));
       return;
     }
 
     return embeddableHandler.render(panelElement, panelState, containerApi)
       .then(embeddable => {
-        return dispatch(embeddableRenderFinished(panelId, embeddable));
+        return dispatch(embeddableRenderFinished(panelState.panelIndex, embeddable));
       })
       .catch(error => {
-        dispatch(embeddableRenderError(panelId, error));
+        dispatch(embeddableRenderError(panelState.panelIndex, error));
       });
   };
 }
diff --git a/src/core_plugins/kibana/public/dashboard/panel/dashboard_panel.js b/src/core_plugins/kibana/public/dashboard/panel/dashboard_panel.js
index 5181086a9..2a614d121 100644
--- a/src/core_plugins/kibana/public/dashboard/panel/dashboard_panel.js
+++ b/src/core_plugins/kibana/public/dashboard/panel/dashboard_panel.js
@@ -7,7 +7,7 @@ import { PanelError } from './panel_error';
 
 export class DashboardPanel extends React.Component {
   async componentDidMount() {
-    this.props.renderEmbeddable(this.panelElement);
+    this.props.renderEmbeddable(this.panelElement, this.props.panel);
   }
 
   toggleExpandedPanel = () => {
diff --git a/src/core_plugins/kibana/public/dashboard/panel/dashboard_panel_container.js b/src/core_plugins/kibana/public/dashboard/panel/dashboard_panel_container.js
index 7cf495a9b..130c95558 100644
--- a/src/core_plugins/kibana/public/dashboard/panel/dashboard_panel_container.js
+++ b/src/core_plugins/kibana/public/dashboard/panel/dashboard_panel_container.js
@@ -19,6 +19,7 @@ import {
   getEmbeddableEditUrl,
   getMaximizedPanelId,
   getEmbeddableError,
+  getPanel
 } from '../reducers';
 
 const mapStateToProps = ({ dashboardState }, { panelId }) => {
@@ -30,11 +31,12 @@ const mapStateToProps = ({ dashboardState }, { panelId }) => {
 
     viewOnlyMode: getFullScreenMode(dashboardState) || getViewMode(dashboardState) === DashboardViewMode.VIEW,
     isExpanded: getMaximizedPanelId(dashboardState) === panelId,
+    panel: getPanel(dashboardState, panelId)
   };
 };
 
 const mapDispatchToProps = (dispatch, { embeddableHandler, panelId, getContainerApi }) => ({
-  renderEmbeddable: (panelElement) => dispatch(renderEmbeddable(embeddableHandler, panelElement, panelId, getContainerApi())),
+  renderEmbeddable: (panelElement, panel) => dispatch(renderEmbeddable(embeddableHandler, panel, panelElement, getContainerApi())),
   onDeletePanel: () => {
     dispatch(deletePanel(panelId));
     dispatch(destroyEmbeddable(panelId, embeddableHandler));
```

(PS! I didn't run this, so I don't actually know if it works — time was running away from me)

Reply:
I can switch to using selectors, though it sounded like there wasn't agreement on whether or not we should consider `getState` an anti-pattern.  I'm not really a fan of passing state through just to get it back in mapDispatchToProps, when the component doesn't actually need it.

Reply:
Ugh, turns out that creates a circular dependency (at least I suspect, though the failure error is slightly vague, it makes sense since now actions requires reducers which requires actions).

/shrug I guess I'll just implement your suggestion for the time being. Though I think it might be a mistake to put selectors in the reducer files because of this.